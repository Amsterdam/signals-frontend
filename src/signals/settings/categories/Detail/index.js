import React, { Fragment, useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams, useLocation } from 'react-router-dom';
import styled from 'styled-components';

import configuration from 'shared/services/configuration/configuration';
import { makeSelectUserCan } from 'containers/App/selectors';
import PageHeader from 'signals/settings/components/PageHeader';
import LoadingIndicator from 'shared/components/LoadingIndicator';

import { fetchCategories } from 'models/categories/actions';
import BackLink from 'components/BackLink';
import routes from 'signals/settings/routes';
import useFetch from 'hooks/useFetch';

import useConfirmedCancel from '../../hooks/useConfirmedCancel';
import useFetchResponseNotification from '../../hooks/useFetchResponseNotification';
import CategoryForm from './components/CategoryForm';

const FormContainer = styled.div`
  // taking into account the space that the FormFooter component takes up
  padding-bottom: 66px;
`;

const getTransformedData = formData => {
  // the API expect a different data structure than we initally received
  // data needs to be transformed before it is sent out ðŸ™„
  const transformedData = { ...formData, new_sla: formData.sla };

  delete transformedData.sla;

  return transformedData;
};

/**
 * Comparison function for incoming and outgoing form data
 *
 * @param   {Object} - First object to compare
 * @param   {Object} - Second object to compare against the first object
 * @returns {Boolean}
 */
const isEqual = ({ description, handling_message, is_active, name, sla }, othValue) =>
  [
    description === othValue.description,
    handling_message === othValue.handling_message,
    is_active === othValue.is_active,
    name === othValue.name,
    sla.n_days === othValue.sla.n_days,
    sla.use_calendar_days === othValue.sla.use_calendar_days,
  ].every(Boolean);

const CategoryDetail = () => {
  const entityName = 'Categorie';

  const location = useLocation();
  const redirectURL = location.referrer || routes.categories;

  const { categoryId } = useParams();
  const isExistingCategory = categoryId !== undefined;

  const { isLoading, isSuccess, error, data, get, patch } = useFetch();
  const historyFetch = useFetch();

  const confirmedCancel = useConfirmedCancel(redirectURL);

  const dispatch = useDispatch();

  const categoryURL = `${configuration.CATEGORIES_PRIVATE_ENDPOINT}${categoryId}`;

  const shouldRenderForm = !isExistingCategory || (isExistingCategory && Boolean(data));

  const userCan = useSelector(makeSelectUserCan);

  const userCanSubmitForm =
    (isExistingCategory && userCan('change_category')) || (!isExistingCategory && userCan('add_category'));


  useFetchResponseNotification({
    entityName,
    error,
    isExisting: isExistingCategory,
    isLoading,
    isSuccess,
    redirectURL,
  });

  const title = `${entityName} ${isExistingCategory ? 'wijzigen' : 'toevoegen'}`;

  const getFormData = useCallback(
    event => {
      const formData = [...new FormData(event.target.form).entries()]
        // convert stringified boolean values to actual booleans
        .map(([key, val]) => [key, key === 'is_active' ? val === 'true' : val])
        // convert line endings
        // by spec, the HTML value should contain \r\n, but the API only contains \n
        .map(([key, val]) => [key, typeof val === 'string' ? val.replace(/\r\n/g, '\n') : val])
        // reduce the entries() array to an object, merging it with the initial data
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), { ...data });

      const sla = {
        n_days: parseInt(formData.n_days, 10),
        use_calendar_days: Boolean(parseInt(formData.use_calendar_days, 10)),
      };

      delete formData.n_days;
      delete formData.use_calendar_days;

      return { ...formData, sla };
    },
    [data]
  );

  const onCancel = useCallback(
    event => {
      const formData = getFormData(event);
      const initialData = Object.entries(data)
        .map(([key, value]) => [key, value === null ? '' : value])
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), { ...data });
      const combinedData = { ...initialData, ...formData };
      const isPristine = isEqual(initialData, combinedData);

      confirmedCancel(isPristine);
    },
    [confirmedCancel, data, getFormData]
  );

  const onSubmit = useCallback(
    event => {
      event.preventDefault();

      const formData = getTransformedData(getFormData(event));

      patch(categoryURL, formData);
    },
    [categoryURL, patch, getFormData]
  );

  useEffect(() => {
    if (isSuccess) {
      dispatch(fetchCategories());
    }
  }, [isSuccess, dispatch]);

  useEffect(() => {
    if (isExistingCategory) {
      get(categoryURL);
      historyFetch.get(`${categoryURL}/history`);
    }

    // Disabling linter; only need to execute on mount; defining the dependencies
    // will throw the component in an endless loop
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Fragment>
      <PageHeader title={title} BackLink={<BackLink to={redirectURL}>Terug naar overzicht</BackLink>} />

      {isLoading && <LoadingIndicator />}

      <FormContainer>
        {shouldRenderForm && (
          <CategoryForm
            data={data}
            history={historyFetch.data}
            onCancel={onCancel}
            onSubmitForm={onSubmit}
            readOnly={!userCanSubmitForm}
          />
        )}
      </FormContainer>
    </Fragment>
  );
};

export default CategoryDetail;
