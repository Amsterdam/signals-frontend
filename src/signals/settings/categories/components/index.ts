export { CategoryForm as default } from './CategoryForm'
export { CategoryDetail } from './Detail'
