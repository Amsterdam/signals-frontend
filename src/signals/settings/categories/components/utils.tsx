// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { FieldNamesMarkedBoolean } from 'react-hook-form'

import type { StandardText } from 'types/api/standard-texts'

import type { CategoryFormPayload, CategoryFormValues } from '../types'
import { Direction } from '../types'

export const getPatchPayload = (
  formData: CategoryFormValues,
  dirtyFields: Partial<Readonly<FieldNamesMarkedBoolean<CategoryFormValues>>>
): CategoryFormPayload => {
  const payload = Object.keys(dirtyFields).map((key) => {
    return {
      [key]: formData[key as keyof CategoryFormValues],
    }
  })

  const payloadObject = Object.assign({}, ...payload)

  const payloadResult = {
    ...payloadObject,
    ...(('n_days' in payloadObject || 'use_calendar_days' in payloadObject) && {
      new_sla: {
        n_days: payloadObject.n_days ?? formData.n_days,
        use_calendar_days: Boolean(Number(payloadObject.use_calendar_days)),
      },
    }),
    ...('show_children_in_filter' in payloadObject && {
      configuration: {
        show_children_in_filter: payloadObject.show_children_in_filter,
      },
    }),
  }

  delete payloadResult['n_days']
  delete payloadResult['use_calendar_days']
  delete payloadResult['show_children_in_filter']
  delete payloadResult['standard_texts']

  return payloadResult
}

export const orderStandardTexts = (
  direction: Direction,
  index: number,
  orderedStandardTexts: StandardText[]
): StandardText[] | undefined => {
  const orderedStandardTextsCopy = [...orderedStandardTexts]
  if (direction === Direction.Up) {
    if (index === 0) {
      return
    }
    const temp = orderedStandardTextsCopy[index - 1]
    orderedStandardTextsCopy[index - 1] = orderedStandardTextsCopy[index]
    orderedStandardTextsCopy[index] = temp
  }
  if (direction === Direction.Down) {
    if (index === orderedStandardTextsCopy.length - 1) {
      return
    }
    const temp = orderedStandardTextsCopy[index + 1]
    orderedStandardTextsCopy[index + 1] = orderedStandardTextsCopy[index]
    orderedStandardTextsCopy[index] = temp
  }
  return orderedStandardTextsCopy
}
