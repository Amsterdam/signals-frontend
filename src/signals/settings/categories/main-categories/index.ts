export { MainCategoriesOverview as default } from './Overview'
