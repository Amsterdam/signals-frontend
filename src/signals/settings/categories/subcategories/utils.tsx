// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type {
  CategoryFormValues,
  CategoryFormPayload,
  DirtyFields,
} from '../types'

export const getPatchPayload = (
  formData: CategoryFormValues,
  dirtyFields: DirtyFields
): CategoryFormPayload => {
  const payload = Object.keys(dirtyFields).map((key) => {
    return {
      [key]: formData[key as keyof CategoryFormValues],
    }
  })

  const payloadObject = Object.assign({}, ...payload)

  if (payloadObject['n_days'] || payloadObject['use_calendar_days']) {
    const updatedPayload = {
      ...payloadObject,
      new_sla: {
        n_days: payloadObject.n_days ?? formData.n_days,
        use_calendar_days: Boolean(Number(payloadObject.use_calendar_days)),
      },
    }

    delete updatedPayload['n_days']
    delete updatedPayload['use_calendar_days']

    return updatedPayload
  } else return payloadObject
}
