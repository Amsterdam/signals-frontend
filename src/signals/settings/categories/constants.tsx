// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Fragment } from 'react'

import { StyledInfo } from './components/styled'

export const ICONEXAMPLE = [
  <Fragment key={'IconExample'}>
    <StyledInfo>Voorbeeld van een icoon:</StyledInfo>
    <img alt={'example-of-an-icon'} src={'/assets/images/afval/rest.svg'} />
  </Fragment>,
]
