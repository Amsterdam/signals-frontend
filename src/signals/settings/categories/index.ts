import { CategoriesOverviewContainer as SubcategoriesOverview } from './subcategories'

export default SubcategoriesOverview
