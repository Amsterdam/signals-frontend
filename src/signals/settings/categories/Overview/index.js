import React, { Fragment, useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Row, Column, themeSpacing } from '@datapunt/asc-ui';
import { useParams, useHistory } from 'react-router-dom';

import PageHeader from 'signals/settings/components/PageHeader';
import LoadingIndicator from 'shared/components/LoadingIndicator';
import Pagination from 'components/Pagination';
import { makeSelectAllSubCategories } from 'models/categories/selectors';
import { makeSelectUserCan } from 'containers/App/selectors';
import { CATEGORY_URL, CATEGORIES_PAGED_URL } from 'signals/settings/routes';
import DataView from 'components/DataView';
import { PAGE_SIZE } from 'containers/App/constants';

import filterData from '../../filterData';

// name mapping from API values to human readable values
export const colMap = {
  fk: 'fk',
  id: 'id',
  value: 'Categorie',
  is_active: 'Status',
  sla: 'Service Level Agreement',
};

const StyledDataView = styled(DataView)`
  th:first-child {
    width: 50%;
  }
`;

const StyledPagination = styled(Pagination)`
  margin-top: ${themeSpacing(12)};
`;

export const CategoriesOverviewContainer = ({ subCategories, userCan }) => {
  const history = useHistory();
  const params = useParams();
  const [page, setPage] = useState(1);

  const pageNum = params.pageNum && parseInt(params.pageNum, 10);
  const count = subCategories && subCategories.length;
  const sliceStart = (pageNum - 1) * PAGE_SIZE;
  const pagedData = (subCategories || [])
    .slice(sliceStart, sliceStart + PAGE_SIZE)
    .map(category => ({
      ...category,
      sla: `${category.sla.n_days} ${
        !category.sla.use_calendar_days ? 'werk' : ''
      }dagen`,
    }));
  const data = filterData(pagedData, colMap);
  const isLoading = !subCategories;


  // subscribe to param changes
  useEffect(() => {
    if (pageNum && pageNum !== page) {
      setPage(pageNum);
    }
  }, [page, pageNum]);

  const onItemClick = useCallback(
    e => {
      if (!userCan('change_category')) {
        e.preventDefault();
        return;
      }

      const {
        currentTarget: {
          dataset: { itemId },
        },
      } = e;

      if (itemId) {
        history.push(`${CATEGORY_URL}/${itemId}`);
      }
    },
    [history, userCan]
  );

  const onPaginationClick = useCallback(
    pageToNavigateTo => {
      global.window.scrollTo(0, 0);
      history.push(`${CATEGORIES_PAGED_URL}/${pageToNavigateTo}`);
    },
    [history]
  );

  const columnHeaders = ['Categorie', 'Service Level Agreement', 'Status'];

  return (
    <Fragment>
      <PageHeader title={`CategorieÃ«n ${count ? `(${count})` : ''}`} />

      <Row>
        {isLoading && <LoadingIndicator />}

        <Column span={12} wrap>
          <Column span={12}>
            <StyledDataView
              headers={columnHeaders}
              columnOrder={columnHeaders}
              onItemClick={onItemClick}
              primaryKeyColumn="fk"
              data={data}
            />
          </Column>

          {!isLoading && count > 0 && (
            <Column span={12}>
              <StyledPagination
                currentPage={page}
                onClick={onPaginationClick}
                totalPages={Math.ceil(count / PAGE_SIZE)}
              />
            </Column>
          )}
        </Column>
      </Row>
    </Fragment>
  );
};

CategoriesOverviewContainer.propTypes = {
  subCategories: PropTypes.arrayOf(PropTypes.shape({})),
  userCan: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  subCategories: makeSelectAllSubCategories,
  userCan: makeSelectUserCan,
});

const withConnect = connect(mapStateToProps);

export default withConnect(CategoriesOverviewContainer);
