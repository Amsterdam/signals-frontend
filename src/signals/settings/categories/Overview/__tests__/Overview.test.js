import React from 'react';
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import * as reactRouterDom from 'react-router-dom';
import { withAppContext } from 'test/utils';

import categories from 'utils/__tests__/fixtures/categories_structured.json';
import { CATEGORY_URL, CATEGORIES_PAGED_URL } from 'signals/settings/routes';
import * as constants from 'containers/App/constants';
import { CategoriesOverviewContainer as CategoriesOverview } from '..';

jest.mock('react-router-dom', () => ({
  __esModule: true,
  ...jest.requireActual('react-router-dom'),
}));

jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
  pageNum: '1',
}));

const push = jest.fn();
jest.spyOn(reactRouterDom, 'useHistory').mockImplementation(() => ({
  push,
}));

jest.mock('containers/App/constants', () => ({
  __esModule: true,
  ...jest.requireActual('containers/App/constants'),
}));

const scrollTo = jest.fn();
global.window.scrollTo = scrollTo;

constants.PAGE_SIZE = 50;

const subCategories = Object.entries(categories).flatMap(([, { sub }]) => sub);
const count = subCategories.length;

describe('signals/settings/categories/containers/Overview', () => {
  beforeEach(() => {
    push.mockReset();
  });

  it('should render header', () => {
    const { getByText, rerender } = render(
      withAppContext(<CategoriesOverview subCategories={null} userCan={() => {}} />)
    );

    expect(getByText('Categorieën')).toBeInTheDocument();

    rerender(withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />));

    expect(getByText(`Categorieën (${count})`)).toBeInTheDocument();
  });

  it('should render paged data', () => {
    const firstCategory = subCategories[0];
    const lastCategory = subCategories[constants.PAGE_SIZE - 1];

    // render the first page
    const { container, rerender } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />)
    );

    expect(container.querySelector(`tr[data-item-id="${firstCategory.fk}"]`)).toBeInTheDocument();
    expect(container.querySelector(`tr[data-item-id="${lastCategory.fk}"]`)).toBeInTheDocument();

    // render the second page
    jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
      pageNum: '2',
    }));

    rerender(withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />));

    const secondPageFirstCategory = subCategories[constants.PAGE_SIZE + 1];
    const secondPageLastCategory = subCategories[constants.PAGE_SIZE * 2 - 1];

    expect(container.querySelector(`tr[data-item-id="${secondPageFirstCategory.fk}"]`)).toBeInTheDocument();
    expect(container.querySelector(`tr[data-item-id="${secondPageLastCategory.fk}"]`)).toBeInTheDocument();
  });

  it('should only render specific data columns', () => {
    const { getByText } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />)
    );

    expect(getByText('Categorie')).toBeInTheDocument();
    expect(getByText('Service Level Agreement')).toBeInTheDocument();
    expect(getByText('Status')).toBeInTheDocument();
  });

  it('should render pagination controls', () => {
    const { rerender, queryByTestId } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />)
    );

    expect(queryByTestId('pagination')).toBeInTheDocument();

    constants.PAGE_SIZE = count + 1;

    rerender(withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />));

    expect(queryByTestId('pagination')).not.toBeInTheDocument();

    constants.PAGE_SIZE = 50;
  });

  it('should push to the history stack and scroll to top on pagination item click', () => {
    jest.spyOn(reactRouterDom, 'useParams').mockImplementation(() => ({
      pageNum: '1',
    }));

    const { getByText } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => {}} />)
    );

    const pageButton = getByText('2');

    expect(scrollTo).not.toHaveBeenCalled();
    expect(push).not.toHaveBeenCalled();

    act(() => {
      fireEvent.click(pageButton);
    });

    expect(push).toHaveBeenCalledTimes(1);
    expect(push).toHaveBeenCalledWith(`${CATEGORIES_PAGED_URL}/2`);
    expect(scrollTo).toHaveBeenCalledWith(0, 0);
  });

  it('should push on list item with an itemId click', async () => {
    const { container } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => true} />)
    );
    const itemId = 666;

    let row;

    await waitFor(() => {
      row = container.querySelector('tbody tr:nth-child(42)');
    });

    // Explicitly set an 'itemId' so that we can easily test against its value.
    row.dataset.itemId = itemId;

    expect(push).toHaveBeenCalledTimes(0);

    act(() => {
      fireEvent.click(row);
    });

    expect(push).toHaveBeenCalledTimes(1);
    expect(push).toHaveBeenCalledWith(`${CATEGORY_URL}/${itemId}`);

    // Remove 'itemId' and fire click event again.
    delete row.dataset.itemId;

    act(() => {
      fireEvent.click(row);
    });

    expect(push).toHaveBeenCalledTimes(1);

    // Set 'itemId' again and fire click event once more.
    row.dataset.itemId = itemId;

    act(() => {
      fireEvent.click(row);
    });

    expect(push).toHaveBeenCalledTimes(2);
  });

  it('should not push on list item click when permissions are insufficient', async () => {
    const { container } = render(
      withAppContext(<CategoriesOverview subCategories={subCategories} userCan={() => false} />)
    );
    const itemId = 666;

    let row;

    await waitFor(() => {
      row = container.querySelector('tbody tr:nth-child(42)');
    });

    // Explicitly set an 'itemId' so that we can easily test against its value.
    row.dataset.itemId = itemId;

    expect(push).not.toHaveBeenCalled();

    act(() => {
      fireEvent.click(row);
    });

    expect(push).not.toHaveBeenCalled();
  });
});
