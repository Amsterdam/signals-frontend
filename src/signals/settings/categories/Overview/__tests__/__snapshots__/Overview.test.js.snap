// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signals/settings/categories/containers/Overview should push to the history stack and scroll to top on pagination item click 1`] = `
<section
  class="PageHeader__StyledSection-sc-1via2ti-0 ceKXIT"
>
  <div
    class="sc-cJSrbW eSjbJi"
  >
    <div>
      <h1
        class="sc-EHOje djuTVt sc-bZQynM hJEdCc PageHeader__StyledHeading-sc-1via2ti-1 iIsPsb"
      >
        CategorieÃ«n (137)
      </h1>
    </div>
  </div>
</section>
`;
